### DISCUSSION TITLE
How to run batches with HypergraphConv

### DISCUSSION THREAD

**Discussion by FernandezR** [pyg-team/pytorch_geometric#7125] [Q&A]
Hello,

I am using the [HypergraphConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.HypergraphConv.html#torch-geometric-nn-conv-hypergraphconv) in my algorithm, but it doesn't seem to handle batch learning. Is there some unique way of doing batches with HypergraphConv that I'm not seeing or is it not possible?

Thanks

**Comment by rusty1s** [comment]
`HypergraphConv` should support batch learning by diagonal stacking hypergraphs. Importantly, if you apply hypergraph learning in a mini-batch scenario, you need to adjust `Data.__inc__` to incorporate for the bipartite incidence matrix. See [here](https://pytorch-geometric.readthedocs.io/en/latest/advanced/batching.html#bipartite-graphs) on how to do so.

**Reply by FernandezR** [reply]
I'm not sure what you mean by diagonal stacking hypergraphs. Can you give an example?

Also, it is not clear how the data should be organized for batching as the implementation doesn't use typical batch handling that is seen in the rest of pytorch. For example, the first dimension is used to get the number of nodes in the graph rather than being used for the batch dimension. 

The scatter ops on the hyperedge index also do not seem to be set up to handle batch computations. From what I can see it expects a 2-dimesional tensor.

**Reply by rusty1s** [reply]
Yes, PyG uses a different mini-batching scheme, since a separate batch dimension does not properly work on graphs with varying number of nodes and edges without inefficient padding. You can read more about this [here](https://pytorch-geometric.readthedocs.io/en/latest/get_started/introduction.html) and [here](https://pytorch-geometric.readthedocs.io/en/latest/advanced/batching.html).

**Reply by FernandezR** [reply]
Can you give an example of how I'm supposed to stack the hypergraphs diagonally? 

Also, the bipartite edge index example doesn't help me with the hypergraph. I don't have source and target nodes so I'm not sure how that is supposed to be modified to work with hypergraphs.

**Reply by FernandezR** [reply]
Also, how do we pass the data loader object to the layers during training? I haven't been able to find an example. 

Is the purpose of the dataloader to merge all the data together into one tensor for data and one tensor for hyperedge index?

**Reply by FernandezR** [reply]
So this is what I need? And then I just pass in the stacked batch tensors created by the loader to the HyGNN layer?

```
class HypergraphData(Data):
    def __inc__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'batch' in key:
            return int(value.max()) + 1
        elif 'index' in key or key == 'face':
            return torch.tensor([[self.num_nodes], [self.num_edges]])
        else:
            return 0
```

**Reply by rusty1s** [reply]
Yes, this looks nearly perfect. Not sure if `self.num_edges` as it probably refers to the number of elements in `edge_index`, not necessarily the number of hyperedges (the second dimension of the incidence matrix).

**Reply by FernandezR** [reply]
Yes, I found that while testing, you need to add a redefinition of the num_edges function. Though, for some reason, the num edges function doesn't carry over to the batch data that is created. Though, it doesn't affect the conversion back to a data list from the batch data object. I can't see how to fix that in the batch data object, as some dynamic inheritance is used to create the batch data object type.

```
class HypergraphData(Data):
    def __inc__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'batch' in key:
            return int(value.max()) + 1
        elif 'index' in key or key == 'face':
            return torch.tensor([[self.num_nodes], [self.num_edges]])
        else:
            return 0

    @property
    def num_edges(self) -> int:
        return int(self.hyperedge_index[1][-1]) + 1
```
